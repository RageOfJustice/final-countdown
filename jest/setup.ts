import 'jest-styled-components'
// тут будут моки библиотек
