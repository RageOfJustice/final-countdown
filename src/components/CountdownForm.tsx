import React, { useState, useRef } from 'react'
import { TextInput, Keyboard } from 'react-native'
import styled from 'src/styled-components'

export type CountdownFormMode = 'start' | 'stop'
export type CountdownFormPressHandler = (minutes?: number) => void

interface Props {
  mode: CountdownFormMode
  onPress?: CountdownFormPressHandler
}

const CountdownForm: React.FC<Props> = ({ mode, onPress }) => {
  const [value, setValue] = useState('')

  const inputRef = useRef<TextInput>(null)

  const isInStartState = mode === 'start'

  const handlePress = () => {
    if (!isInStartState) {
      onPress?.()
    }
    const parsed = parseInt(value, 10)
    if (!isNaN(parsed)) {
      onPress?.(parsed)
    }
    setValue('')
    inputRef.current?.clear()
    Keyboard.dismiss()
  }

  return (
    <Wrapper>
      <Text>Countdown:</Text>
      <Input
        ref={inputRef}
        value={value}
        editable={isInStartState}
        onChangeText={setValue}
        placeholder="Minutes"
        keyboardType="decimal-pad"
      />
      <ButtonWrapper
        disabled={isInStartState && !value}
        onPress={handlePress}
        testID="controlButton"
      >
        <ButtonText>{isInStartState ? 'Start' : ' Stop'}</ButtonText>
      </ButtonWrapper>
    </Wrapper>
  )
}

const Wrapper = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: center;
`

const Text = styled.Text`
  color: ${({ theme }) => theme.colors.black};
`

const Input = styled.TextInput.attrs(({ theme }) => ({
  underlineColorAndroid: 'transparent',
  placeholderTextColor: theme.colors.lightGrey,
  selectionColor: theme.colors.darkGrey,
}))`
  color: ${({ theme }) => theme.colors.black};
  margin-horizontal: 8px;
  border-width: 1px;
  padding: 8px 4px;
  width: 100px;
  border-color: ${({ theme }) => theme.colors.black};
  opacity: ${({ editable }) => (editable ? 1 : 0.6)};
`

const ButtonWrapper = styled.TouchableOpacity`
  padding: 8px;
  opacity: ${({ disabled }) => (disabled ? 0.6 : 1)};
  background-color: ${({ theme }) => theme.colors.cyan};
`

const ButtonText = styled.Text`
  color: ${({ theme }) => theme.colors.white};
`

export default CountdownForm
