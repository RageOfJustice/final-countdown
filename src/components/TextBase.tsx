import { Platform } from 'react-native'

import styled from 'src/styled-components'

// на некоторых андроидах обрезаются слова
// поэтому надо явно указать шрифт
const TextBase = styled.Text`
  ${Platform.OS === 'android' && 'font-family: Roboto;'}
`

export default TextBase
