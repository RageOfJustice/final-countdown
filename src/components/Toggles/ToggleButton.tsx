import React from 'react'
import styled from 'src/styled-components'

interface Props {
  isActive?: boolean
  onPress?: () => void
  title: string
}

const ToggleButton: React.FC<Props> = ({ isActive, title, onPress }) => (
  <Wrapper onPress={onPress} isActive={isActive}>
    <Text isActive={isActive}>{title}</Text>
  </Wrapper>
)

const Wrapper = styled.TouchableOpacity<{ isActive?: boolean }>`
  padding: 8px 16px;
  border-width: 1px;
  border-color: ${({ theme }) => theme.colors.black};
  background-color: ${({ theme, isActive }) =>
    isActive ? theme.colors.darkGrey : 'transparent'};
  align-items: center;
  justify-content: center;
`

const Text = styled.Text<{ isActive?: boolean }>`
  color: ${({ isActive, theme }) =>
    isActive ? theme.colors.white : theme.colors.black};
  font-size: 14px;
`

export default ToggleButton
