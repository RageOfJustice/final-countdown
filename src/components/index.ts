export * from './Timer'
export * from './Toggles'
export * from './CountdownForm'

export { default as Timer } from './Timer'
export { default as Toggles } from './Toggles'
export { default as CountdownForm } from './CountdownForm'
