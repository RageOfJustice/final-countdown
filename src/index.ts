import React from 'react'

interface Props {}

const App: React.FC<Props> = props => {
  return null
}

export default App
