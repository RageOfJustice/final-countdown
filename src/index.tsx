import React from 'react'
import { ThemeProvider } from './styled-components'
import theme from './theme'

interface Props {}

const App: React.FC<Props> = () => {
  return <ThemeProvider theme={theme} />
}

export default App
