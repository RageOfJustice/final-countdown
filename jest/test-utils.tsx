import React, { ReactElement, FC } from 'react'
import { render, RenderOptions } from '@testing-library/react-native'
import theme from 'src/theme'
import { ThemeProvider } from 'src/styled-components'

const Providers: FC = ({ children }) => {
  return <ThemeProvider theme={theme}>{children as ReactElement}</ThemeProvider>
}

const customRender = (
  ui: ReactElement<any>,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react-native'

export { customRender as render }
