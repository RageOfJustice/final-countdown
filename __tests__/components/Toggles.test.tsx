import React from 'react'
import { render, fireEvent } from '../../jest/test-utils'
import { Toggles } from 'src/components'
// тут такой импорт, потому что кнопка "скрыта" от отсальной части приложения
// является частью компонента Toggles
import ToggleButton from 'src/components/Toggles/ToggleButton'

describe('ToggleButton', () => {
  it('renders correclty', () => {
    const { container } = render(<ToggleButton title="title" />)

    expect(container.children[0]).toMatchSnapshot()
  })
  it('renders correclty in active state', () => {
    const { container } = render(<ToggleButton isActive title="title" />)

    expect(container.children[0]).toMatchSnapshot()
  })
  it('can be pressed', () => {
    const onPress = jest.fn()
    const { container } = render(
      <ToggleButton onPress={onPress} title="title" />,
    )
    fireEvent.press(container.children[0])
    expect(onPress).toHaveBeenCalled()
  })
})

describe('Toggles', () => {
  const values = [
    { title: '1X', value: 1 },
    { title: '1.5X', value: 1.5 },
    { title: '2X', value: 2 },
  ]

  it('renders correctly', () => {
    const { container } = render(<Toggles values={values} />)

    expect(container.children[0]).toMatchSnapshot()
  })

  it('can be pressed', () => {
    const onPress = jest.fn()
    const { getAllByTestId } = render(
      <Toggles onPress={onPress} values={values} />,
    )
    const component = getAllByTestId('toggleButton')[1]
    fireEvent.press(component)
    expect(onPress).toHaveBeenCalledWith(values[1], 1)
  })
})
