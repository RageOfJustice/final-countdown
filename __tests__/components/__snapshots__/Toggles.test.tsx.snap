// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleButton renders correclty 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderColor": "#2d3436",
        "borderWidth": 1,
        "justifyContent": "center",
        "paddingBottom": 8,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 8,
      },
    ]
  }
  testID="toggleButton"
>
  <Text
    style={
      Array [
        Object {
          "color": "#2d3436",
          "fontSize": 14,
        },
      ]
    }
  >
    title
  </Text>
</TouchableOpacity>
`;

exports[`ToggleButton renders correclty in active state 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  isActive={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#636e72",
        "borderColor": "#2d3436",
        "borderWidth": 1,
        "justifyContent": "center",
        "paddingBottom": 8,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 8,
      },
    ]
  }
  testID="toggleButton"
>
  <Text
    isActive={true}
    style={
      Array [
        Object {
          "color": "#FFFFFF",
          "fontSize": 14,
        },
      ]
    }
  >
    title
  </Text>
</TouchableOpacity>
`;

exports[`Toggles renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
        "marginHorizontal": -2,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "paddingHorizontal": 2,
        },
      ]
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      isActive={true}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#636e72",
            "borderColor": "#2d3436",
            "borderWidth": 1,
            "justifyContent": "center",
            "paddingBottom": 8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 8,
          },
        ]
      }
      testID="toggleButton"
    >
      <Text
        isActive={true}
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontSize": 14,
            },
          ]
        }
      >
        1X
      </Text>
    </TouchableOpacity>
  </View>
  <View
    style={
      Array [
        Object {
          "paddingHorizontal": 2,
        },
      ]
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      isActive={false}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#2d3436",
            "borderWidth": 1,
            "justifyContent": "center",
            "paddingBottom": 8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 8,
          },
        ]
      }
      testID="toggleButton"
    >
      <Text
        isActive={false}
        style={
          Array [
            Object {
              "color": "#2d3436",
              "fontSize": 14,
            },
          ]
        }
      >
        1.5X
      </Text>
    </TouchableOpacity>
  </View>
  <View
    style={
      Array [
        Object {
          "paddingHorizontal": 2,
        },
      ]
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      isActive={false}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#2d3436",
            "borderWidth": 1,
            "justifyContent": "center",
            "paddingBottom": 8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 8,
          },
        ]
      }
      testID="toggleButton"
    >
      <Text
        isActive={false}
        style={
          Array [
            Object {
              "color": "#2d3436",
              "fontSize": 14,
            },
          ]
        }
      >
        2X
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;
